package/MDAnalysis/coordinates/chemfiles.py

```diff
@@ -0,0 +1,335 @@
+        self._cached_n_atoms = None
+        self._open()
+        self.ts = self._Timestep(self.n_atoms, **self._ts_kwargs)
+        self.next()
+
+    def _open(self):
+        self._closed = False
+        self._step = 0
+        self._frame = -1
+        # Open file last to ensure that all other attributes are set
+        # in case of exception
+        self._file = Trajectory(self.filename, 'r', self._format)
+
+    def close(self):
+        """close reader"""
+        if not self._closed:
+            self._file.close()
+            self._closed = True
+
+    @property
+    def n_frames(self):
+        """number of frames in trajectory"""
+        return self._file.nsteps
+
+    @property
+    def n_atoms(self):
+        """number of atoms in the first frame of the trajectory"""
+        if self._cached_n_atoms is None:
+            self._cached_n_atoms = len(self._file.read_step(0).atoms)
+        return self._cached_n_atoms
+
+    def _reopen(self):
+        """reopen trajectory"""
+        self.close()
+        self._open()
+
+    def _read_frame(self, i):
+        """read frame i"""
+        self._step = i
+        return self._read_next_timestep()
+
+    def _read_next_timestep(self, ts=None):
+        """copy next frame into timestep"""
+        if self._step >= self.n_frames:
+            raise IOError('trying to go over trajectory limit')
+        if ts is None:
+            ts = self.ts
+        self.ts = ts
+        frame = self._file.read_step(self._step)
+        self._frame_to_ts(frame, ts)
+        ts.frame = frame.step
+        self._step += 1
+        return ts
+
+    def _frame_to_ts(self, frame, ts):
+        """convert a chemfiles frame to a :class:`TimeStep`"""
+        if len(frame.atoms) != self.n_atoms:
+            raise IOError(
+                "The number of atom changed in the trajectory. "
+                "This is not supported by MDAnalysis."
+            )
+
+        ts.dimensions[:] = frame.cell.lengths + frame.cell.angles
+        ts.positions[:] = frame.positions[:]
+        if frame.has_velocities():
+            ts.has_velocities = True
+            ts.velocities[:] = frame.velocities[:]
+
+    def Writer(self, filename, n_atoms=None, **kwargs):
+        """Return writer for trajectory format"""
+        if n_atoms is None:
+            n_atoms = self.n_atoms
+        return ChemfilesWriter(filename, n_atoms, **kwargs)
+
+
+class ChemfilesWriter(base.WriterBase):
+    """
+    Coordinate writer using chemfiles.
+
+    The file format to used is guessed based on the file extension. If no
+    matching format is found, a ``ChemfilesError`` is raised. It is also
+    possible to manually specify the format to use for a given file.
+
+    Chemfiles support writting to files with varying number of atoms if the
+    underlying format support it. This is the case of most of text-based
+    formats.
+    """
+
+    format = 'chemfiles'
+    multiframe = True
+    units = {'time': 'fs', 'length': 'Angstrom'}
```

Comment:
This can be confusing for users. As I understand chemfiles can write many different file formats. The units will vary between formats. So writing an XTC file with this could lead to wrong results.

Reply:
does the `units` declare what the writer wants the data to be in, or how it will write the data?

Reply:
Both actually. The [iobase](https://github.com/MDAnalysis/mdanalysis/blob/abc90493f27ad160e311405f78d4377ac7a357ef/package/MDAnalysis/coordinates/base.py#L1107) converts the units during reading into a common unit (I think time:ns and length:Angstrom) and converts it back into native format during writing. It helps for users if units in MDAnalysis are independent of the filetype of the trajectory.

Reply:
So chemfiles mostly assumes the above units for the memory representation, and then convert into these when reading, and into the file format unit when writing. So passing an atom at [10, 12, 33] to chemfiles and writing it to XTC would write coordinates as (1, 1.2, 3.3).

The mostly part is because some format don't have a specified unit (XYZ for example), so we just assume they are in A and fs.

Reply:
OK this should work then. I would leave a comment like this how chemfiles handles units so others knows this works for XTC files as well.